### PR TITLE
Hide Stories options and add deprecation notice

### LINKED ISSUE #4201: Hide experience options if no Stories are created, to replace with notice about upcoming Stories plugin
## Feature description

The AMP Stories editor is no longer compatible with the latest Gutenberg version (see #4200). For this reason and because work is underway on a new effort to create a standalone Stories plugin (see #4172), we need to sunset the AMP Stories editor in the AMP plugin.

The first part of this, as noted in #4074, we need to hide the Stories experience entirely (the experience checkboxes) if the user has no `amp_story` posts created. The thinking being that we can continue to allow access to Stories for users who are using them, but we discontinue offering the Stories feature for new installs or people who don't have stories created. In the place of the experience checkboxes, there can be a notice that says a new Stories plugin is coming and link to it when it is available for testing.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### LINKED ISSUE #4202: Add deprecation and migration/export notice to users of Stories
## Feature description

Work is underway on a new effort to create a standalone Stories plugin (see #4172), we need to sunset the AMP Stories editor in the AMP plugin. As part of this, we need to notify users as soon as possible that they need to backup or export stories they have created since they are not likely to be automatically migrated to the [new standalone plugin](https://github.com/google/web-stories-wp/).

This notice would only appear for users who (1) have the Stories experience enabled and (2) have existing `amp_story` posts created. For other users, the experiences would be hidden in favor of a single notice on the AMP settings screen per #4201.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by pierlon**
### Stories experience enabled

The following table summarizes what happens when the Stories experience is enabled:

‏‏‎ ‎ | With `amp_story` posts | Without `amp_story` posts
---|:---:|:---:
Story posts can be managed | :heavy_check_mark:  | :x:
Stories settings are available | :heavy_check_mark: | :x:
Notice shown on AMP settings page | image | image

NB: In the case with existing `amp_story` posts, the deprecated notice is also shown on the Stories post list page:

image

And also in the Story editor:

image

And an admin pointer on non-story screens, when the Stories experience is enabled:

Screen Shot 2020-02-08 at 21 16 58

### Stories experience disabled

The following table summarizes what happens when the Stories experience is disabled:

‏‏‎ ‎ | With `amp_story` posts | Without `amp_story` posts
---|:---:|:---:
Story posts can be managed | :x:  | :x:
Stories settings are available | :heavy_check_mark: | :x:
Notice shown on AMP settings page | image | image

---

Fixes #4201.
Fixes #4202.

## Checklist

- [x] My pull request is addressing an [open issue](https://github.com/ampproject/amp-wp/contributing/project-management.md#life-of-an-issue) (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/contributing/engineering.md#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/contributing/engineering.md) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: includes/options/class-amp-options-manager.php [hunk 1]
```diff
@@ -562,6 +573,20 @@ public static function persistent_object_caching_notice() {
...
+	/**
+	 * Render the Stories deprecation admin notice.
+	 */
+	public static function render_stories_deprecation_notice() {
+		if ( 'toplevel_page_' . self::OPTION_NAME !== get_current_screen()->id ) {
+			return;
+		}
+
+		printf(
+			'<div class="notice notice-error"><p>%s</p></div>',
```
Comment (westonruter) [inline includes/options/class-amp-options-manager.php]:
```suggestion
			'<div class="notice notice-warn"><p>%s</p></div>',
```

File: includes/options/class-amp-options-manager.php [hunk 2]
```diff
@@ -562,6 +573,20 @@ public static function persistent_object_caching_notice() {
...
+			esc_html__( 'AMP Stories has been deprecated and will no longer be supported. A separate plugin will soon be available for testing.', 'amp' )
```
Comment (westonruter) [inline includes/options/class-amp-options-manager.php]:
```suggestion
			esc_html__( 'The Stories experience in the AMP plugin has been deprecated and will no longer be supported. A separate plugin for Stories be available soon for testing. Please backup or export your existing Stories as they will not be available in the next version of the AMP plugin.', 'amp' )
```

File: includes/options/class-amp-options-manager.php [hunk 3]
```diff
@@ -199,12 +200,22 @@ public static function is_website_experience_enabled() {
...
+	 * @param bool $include_posts_check Include check to see if any `amp_story` posts exists.
 	 * @return bool Enabled.
 	 */
-	public static function is_stories_experience_enabled() {
+	public static function is_stories_experience_enabled( $include_posts_check = true ) {
+		$has_stories_posts = true;
+
+		// Note: A post count cannot be done until the post type is registered (see AMP_Story_Post_Type::register).
+		if ( $include_posts_check ) {
+			$story_posts_count       = (array) wp_count_posts( AMP_Story_Post_Type::POST_TYPE_SLUG );
```
Comment (westonruter) [inline includes/options/class-amp-options-manager.php]: This doesn't seem quite right. I have a draft Story post and it is is not getting picked up here.

Comment (westonruter) [reply includes/options/class-amp-options-manager.php]: Actually, this is not working for me even for published `amp_story` posts.

**Comment by kienstra** [comment]
It looks like this also fixes #4202, right?

>As part of this, we need to notify users as soon as possible that they need to backup or export stories they have created since they are not likely to be automatically migrated to the new standalone plugin.

**Comment by pierlon** [comment]
Nope. The notice here is only applied to the AMP settings screen and is shown irregardless of Stories being enabled or not.

Should this PR also fix #4202, @westonruter?

File: includes/options/class-amp-options-manager.php (hunk 3 above)
Comment (pierlon) [reply includes/options/class-amp-options-manager.php]: Fixed in https://github.com/ampproject/amp-wp/pull/4219/commits/e1aece941277e5d3a2f2c3ced1950c56b6e6609a.

**Comment by kienstra** [comment]
Ah, if this PR doesn't fix #4202, I don't think this PR needs to fix it.

**Comment by westonruter** [comment]
The entire experiences selection should be omitted entirely now:

image

So this code should not be run if Stories is not available:

https://github.com/ampproject/amp-wp/blob/19d54bdf0ac4c41a37927e11241b723cdb754de8/includes/options/class-amp-options-menu.php#L82-L91

**Comment by westonruter** [comment]
> So this code should not be run if Stories is not available:

I mean, if no `amp_story` posts exist, then the experiences section checkbox should be hidden.

**Comment by westonruter** [comment]
> Should this PR also fix #4202

Yes.

**Comment by westonruter** [comment]
So this PR can also fix #4203.

**Comment by westonruter** [comment]
Rather, the code should not yet be removed, but the experiences can be omitted. Issue #4203 would need to be fixed in a subsequent PR which removes all stories code.

File: includes/options/class-amp-options-manager.php [hunk 4]
```diff
@@ -562,6 +567,20 @@ public static function persistent_object_caching_notice() {
...
+	 */
+	public static function render_stories_deprecation_notice() {
+		if ( 'toplevel_page_' . self::OPTION_NAME !== get_current_screen()->id ) {
+			return;
+		}
+
+		printf(
+			'<div class="notice notice-warning"><p>%s</p></div>',
+			esc_html__( 'The Stories experience in the AMP plugin has been deprecated and will no longer be supported. A separate plugin for Stories be available soon for testing. Please backup or export your existing Stories as they will not be available in the next version of the AMP plugin.', 'amp' )
+		);
```
Comment (westonruter) [inline includes/options/class-amp-options-manager.php]: If the Stories experience is hidden, then instead of a warning there should be info notice to say something like:

> The Stories experience has been removed from the AMP plugin. This beta feature is being split into a separate standalone plugin which will be available for installation soon.

Comment (pierlon) [reply includes/options/class-amp-options-manager.php]: Done in https://github.com/ampproject/amp-wp/pull/4219/commits/2dd83a06d081a03f32bfd2532b9706d53a52426e.

File: includes/options/class-amp-options-manager.php [hunk 5]
```diff
@@ -562,6 +567,23 @@ public static function persistent_object_caching_notice() {
...
 
+	/**
+	 * Render the Stories deprecation admin notice.
+	 */
+	public static function render_stories_deprecation_notice() {
+		if ( self::is_stories_experience_enabled() ) {
+			printf(
+				'<div class="notice notice-warning"><p>%s</p></div>',
+				esc_html__( 'The Stories experience in the AMP plugin has been deprecated and will no longer be supported. A separate plugin for Stories be available soon for testing. Please backup or export your existing Stories as they will not be available in the next version of the AMP plugin.', 'amp' )
+			);
```
Comment (westonruter) [inline includes/options/class-amp-options-manager.php]: This should be shown on the AMP admin settings screen as well as the Stories post list admin screen. Otherwise, it should not be shown, since it would be excessively spammy I think.

**Comment by pierlon** [comment]
I've also added the warning notice to the Story editor:

image

File: includes/options/class-amp-options-manager.php (hunk 5 above)
Comment (pierlon) [reply includes/options/class-amp-options-manager.php]: Done in https://github.com/ampproject/amp-wp/pull/4219/commits/ef30768fc89d12d477d27bf7a8ab6dd1e7f9cecc.

File: includes/options/class-amp-options-manager.php (hunk 2 above)
Comment (amedina) [reply includes/options/class-amp-options-manager.php]: Should we point the user to instructions on how to do the back up? (e.g. using the exporting capability)

File: includes/options/class-amp-options-manager.php [hunk 6]
```diff
@@ -562,6 +571,47 @@ public static function persistent_object_caching_notice() {
...
+		if (
+			self::is_stories_experience_enabled() &&
+			(
+				'edit-amp_story' === get_current_screen()->id ||
+				'toplevel_page_' . self::OPTION_NAME === get_current_screen()->id
+			)
+		) {
+			printf(
+				'<div class="notice notice-warning"><p>%s</p></div>',
+				esc_html__( 'The Stories experience in the AMP plugin has been deprecated and will no longer be supported. A separate plugin for Stories be available soon for testing. Please backup or export your existing Stories as they will not be available in the next version of the AMP plugin.', 'amp' )
```
Comment (westonruter) [inline includes/options/class-amp-options-manager.php]:
```suggestion
				esc_html__( 'The Stories experience in the AMP plugin has been deprecated and will no longer be supported. A separate plugin for Stories will be available soon for testing. Please backup or export your existing Stories as they will not be available in the next version of the AMP plugin.', 'amp' )
```

File: includes/options/class-amp-options-manager.php [hunk 7]
```diff
@@ -562,6 +571,47 @@ public static function persistent_object_caching_notice() {
...
+	public static function render_stories_deprecation_editor_notice() {
+		$script = sprintf(
+			"( function( wp ) {
+						wp.data.dispatch( 'core/notices' ).createNotice(
+							'warning',
+							'%s',
+							{ isDismissible: false }
+						);
+					} )( window.wp );",
+			esc_html__( 'The Stories experience in the AMP plugin has been deprecated and will no longer be supported. A separate plugin for Stories be available soon for testing. Please backup or export your existing Stories as they will not be available in the next version of the AMP plugin.', 'amp' )
```
Comment (westonruter) [inline includes/options/class-amp-options-manager.php]:
```suggestion
			esc_html__( 'The Stories experience is being extracted from the AMP plugin into a separate standalone plugin which will be available soon. Please backup or export your existing Stories as they will not be available in the next version of the AMP plugin.', 'amp' )
```

Comment (westonruter) [inline includes/options/class-amp-options-manager.php]: Also, instead of `'%s'` this should be just `%s` and instead of `esc_html__()` we need to use `wp_json_encode( __( ... ) )`

File: includes/options/class-amp-options-manager.php [hunk 8]
```diff
@@ -66,10 +66,15 @@ public static function register_settings() {
...
 		add_action( 'update_option_' . self::OPTION_NAME, [ __CLASS__, 'maybe_flush_rewrite_rules' ], 10, 2 );
 		add_action( 'admin_notices', [ __CLASS__, 'render_welcome_notice' ] );
+		add_action( 'admin_notices', [ __CLASS__, 'render_stories_deprecation_notice' ] );
 		add_action( 'admin_notices', [ __CLASS__, 'persistent_object_caching_notice' ] );
 		add_action( 'admin_notices', [ __CLASS__, 'render_cache_miss_notice' ] );
 		add_action( 'admin_notices', [ __CLASS__, 'render_php_css_parser_conflict_notice' ] );
 		add_action( 'admin_notices', [ __CLASS__, 'insecure_connection_notice' ] );
+
+		if ( self::is_stories_experience_enabled() ) {
+			add_action( 'enqueue_block_editor_assets', [ __CLASS__, 'render_stories_deprecation_editor_notice' ] );
```
Comment (westonruter) [inline includes/options/class-amp-options-manager.php]: This is a great touch.

File: includes/options/class-amp-options-manager.php (hunk 7 above)
Comment (pierlon) [reply includes/options/class-amp-options-manager.php]: Done in https://github.com/ampproject/amp-wp/pull/4219/commits/3e55b5ed4d6b2d8e42e0072636de0b717fe8c9ca.

File: includes/options/class-amp-options-manager.php (hunk 8 above)
Comment (pierlon) [reply includes/options/class-amp-options-manager.php]: Gracias.

**Comment by westonruter** [comment]
Please also link the Story export text to https://amp-wp.org/documentation/amp-stories/exporting-stories/

File: includes/options/class-amp-options-manager.php [hunk 9]
```diff
@@ -603,7 +611,15 @@ public static function render_stories_deprecation_editor_notice() {
 						wp.data.dispatch( 'core/notices' ).createNotice(
 							'warning',
 							%s,
-							{ isDismissible: false }
+							{
+						  		isDismissible: false,
+						  		actions: [
+									{
+										url: 'https://amp-wp.org/documentation/amp-stories/exporting-stories/',
```
Comment (pierlon) [inline includes/options/class-amp-options-manager.php]: (Apparently) links cannot be added inline, so I added the link to exporting Stories as an action.

File: includes/options/class-amp-options-manager.php [hunk 10]
```diff
@@ -562,6 +571,63 @@ public static function persistent_object_caching_notice() {
...
+			"( function( wp ) {
+						wp.data.dispatch( 'core/notices' ).createNotice(
+							'warning',
+							%s,
+							{
+						  		isDismissible: false,
+						  		actions: [
+									{
+										url: 'https://amp-wp.org/documentation/amp-stories/exporting-stories/',
+										label: 'View how to export your Stories',
```
Comment (westonruter) [inline includes/options/class-amp-options-manager.php]: This needs to be translatable.

```suggestion
										label: <?php echo wp_json_encode( __( 'View how to export your Stories', 'amp' ) ) ?>,
```

File: includes/options/class-amp-options-manager.php [hunk 11]
```diff
@@ -562,6 +571,63 @@ public static function persistent_object_caching_notice() {
...
+				'%s %s %s',
+				esc_html__( 'The Stories experience is being extracted from the AMP plugin into a separate standalone plugin which will be available soon. Please backup or', 'amp' ),
+				sprintf(
+					'<a href="%s">%s</a>',
+					esc_url( 'https://amp-wp.org/documentation/amp-stories/exporting-stories/' ),
+					esc_html__( 'export your existing Stories', 'amp' )
+				),
+				esc_html__( 'as they will not be available in the next version of the AMP plugin.', 'amp' )
+			);
+			echo '</p></div>';
```
Comment (westonruter) [inline includes/options/class-amp-options-manager.php]: Please refactor the warning to match the warning in the editor:

image

This will reduce the strings to translate.

Aside: Breaking up the strings in this way would make it hard to translate in some languages.

**Comment by pierlon** [comment]
:warning: Rebasing with latest changes from `develop`.

File: includes/options/class-amp-options-manager.php (hunk 11 above)
Comment (pierlon) [reply includes/options/class-amp-options-manager.php]: Good point, made the change in 8c14de5c2a84021d6921960a70039c109a7f4398.

**Comment by pierlon** [comment]
The following can be observed when the Stories experience is disabled while editing a Story and refreshing the page:

image

Along with the following stack-trace:

```
[08-Feb-2020 23:55:26 UTC] PHP Notice:  map_meta_cap was called <strong>incorrectly</strong>. The post type amp_story is not registered, so it may not be reliable to check the capability "edit_post" against a post of that type. Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 4.4.0.) in /app/public/core-dev/src/wp-includes/functions.php on line 5166
[08-Feb-2020 23:55:26 UTC] PHP Stack trace:
[08-Feb-2020 23:55:26 UTC] PHP   1. {main}() /app/public/core-dev/src/wp-admin/admin-ajax.php:0
[08-Feb-2020 23:55:26 UTC] PHP   2. do_action() /app/public/core-dev/src/wp-admin/admin-ajax.php:175
[08-Feb-2020 23:55:26 UTC] PHP   3. WP_Hook->do_action() /app/public/core-dev/src/wp-includes/plugin.php:478
[08-Feb-2020 23:55:26 UTC] PHP   4. WP_Hook->apply_filters() /app/public/core-dev/src/wp-includes/class-wp-hook.php:311
[08-Feb-2020 23:55:26 UTC] PHP   5. wp_ajax_wp_remove_post_lock() /app/public/core-dev/src/wp-includes/class-wp-hook.php:287
[08-Feb-2020 23:55:26 UTC] PHP   6. current_user_can() /app/public/core-dev/src/wp-admin/includes/ajax-actions.php:2816
[08-Feb-2020 23:55:26 UTC] PHP   7. WP_User->has_cap() /app/public/core-dev/src/wp-includes/capabilities.php:668
[08-Feb-2020 23:55:26 UTC] PHP   8. map_meta_cap() /app/public/core-dev/src/wp-includes/class-wp-user.php:749
[08-Feb-2020 23:55:26 UTC] PHP   9. _doing_it_wrong() /app/public/core-dev/src/wp-includes/capabilities.php:161
[08-Feb-2020 23:55:26 UTC] PHP  10. trigger_error() /app/public/core-dev/src/wp-includes/functions.php:5166
```
I don't think this needs to be looked into since it would be such a rare edge case, but just want it to be known.

**Comment by pierlon** [comment]
@westonruter how should the [failing E2E tests](https://travis-ci.org/ampproject/amp-wp/jobs/647883806#L960-L1014) be handled? All are related to the experiences section not being shown.

**Comment by westonruter** [comment]
@pierlon Just remove the offending tests. Everything will be removed with #4203 anyway.

**Comment by westonruter** [comment]
One minor thing that I found is that if I trash all the stories and then empty the trash, then I get `wp_die()` message saying the post type is invalid. This is because once there are no posts left, the post type then gets unregistered. I suppose that is fine, as most users will not try to empty their trash. And if they do, they can just go back to the WP Admin.
